File: node/src/sensor/illuminance.ts
import { Sensor, SensorType, SensorCategory } from './base.js';
import { defineSensor, SensorDomain } from './meta.js';

import type { Observable } from '../observable/index.js';
import type { PropertyChangeOf, SensorLike, SensorOptions } from './base.js';

/**
 * Property names of an illuminance sensor.
 *
 * @internal
 */
export enum IlluminanceProperty {
  /** Current illuminance in lux. */
  Current = 'current',
}

/**
 * Property values of an illuminance sensor.
 *
 * @internal
 */
export interface IlluminanceInfoProperties {
  [IlluminanceProperty.Current]: number;
}

/** Read-only proxy interface for an illuminance sensor. */
export interface IlluminanceInfoLike extends SensorLike {
  readonly type: SensorType.Illuminance;
  readonly onPropertyChanged: Observable<PropertyChangeOf<IlluminanceInfoProperties>>;

  getValue(property: IlluminanceProperty.Current): number | undefined;
  getValue(property: string): unknown;
}

/** Illuminance info sensor. Reports current light level in lux. */
export class IlluminanceInfo<TStorage extends object = Record<string, any>> extends Sensor<IlluminanceInfoProperties, TStorage> {
  readonly type = SensorType.Illuminance;
  readonly category = SensorCategory.Info;

  constructor(name = 'Illuminance', options?: SensorOptions) {
    super(name, options);

    this._writeState({ [IlluminanceProperty.Current]: 0 });
  }

  get current(): number {
    return this.props.current;
  }

  /**
   * Report a new illuminance reading. Clamped to [0, 200000] lx.
   *
   * @param value - Illuminance reading in lux.
   *
   * @example
   * ```ts
   * illuminance.setCurrent(120);
   * ```
   */
  setCurrent(value: number): void {
    this._writeState({ [IlluminanceProperty.Current]: Math.max(0, Math.min(200000, value)) });
  }

  /**
   * Read-only sensor: external writes are ignored.
   *
   * @internal
   */
  updateValue(_property: string, _value: unknown): void {}
}

/** Registry metadata for {@link IlluminanceInfo}. */
export const illuminanceMeta = defineSensor({
  type: SensorType.Illuminance,
  category: SensorCategory.Info,
  assignmentKey: 'illuminance',
  multiProvider: true,
  isDetectionType: false,
  properties: {
    [IlluminanceProperty.Current]: { type: 'number', unit: 'lx', writable: true },
  },
  shortcutable: true,
  virtual: { properties: { [IlluminanceProperty.Current]: 20 } },
  semantics: {
    domain: SensorDomain.Measurement,
    stateProperty: IlluminanceProperty.Current,
    commandProperty: IlluminanceProperty.Current,
    deviceClass: 'illuminance',
    unit: 'lx',
  },
});

File: python/camera_ui_sdk/sensor/illuminance.py
from __future__ import annotations

from collections.abc import Mapping
from enum import StrEnum
from typing import Any, Generic, Literal, Protocol, overload, runtime_checkable

from typing_extensions import TypedDict, TypeVar

from ..observable import Observable
from .base import Sensor, SensorCategory, SensorLike, SensorType


class IlluminanceProperty(StrEnum):
    """Property names of an illuminance sensor."""

    Current = "current"
    """Current illuminance in lux."""


class IlluminanceInfoProperties(TypedDict):
    """Property values of an illuminance sensor."""

    current: float


class IlluminancePropertyChangeData(TypedDict):
    """Property change payload emitted on IlluminanceInfoLike.onPropertyChanged."""

    property: str
    """Name of the changed property, a IlluminanceProperty value."""
    value: float
    """New value of the property."""


TStorage = TypeVar("TStorage", bound=Mapping[str, Any], default=dict[str, Any])


@runtime_checkable
class IlluminanceInfoLike(SensorLike, Protocol):
    """Read-only proxy interface for an illuminance sensor."""

    @property
    def type(self) -> SensorType:
        return SensorType.Illuminance

    @property
    def onPropertyChanged(self) -> Observable[IlluminancePropertyChangeData]: ...

    @overload
    def getValue(self, property: Literal[IlluminanceProperty.Current]) -> float | None: ...
    @overload
    def getValue(self, property: str) -> object | None: ...


class IlluminanceInfo(Sensor[IlluminanceInfoProperties, TStorage, str], Generic[TStorage]):
    """Illuminance info sensor. Reports current light level in lux."""

    _requires_frames = False

    def __init__(self, name: str = "Illuminance", *, native_id: str | None = None) -> None:
        super().__init__(name, native_id=native_id)
        self._write_state({IlluminanceProperty.Current.value: 0})

    @property
    def type(self) -> SensorType:
        return SensorType.Illuminance

    @property
    def category(self) -> SensorCategory:
        return SensorCategory.Info

    @property
    def current(self) -> float:
        value = self.props.current
        return float(value) if value is not None else 0.0

    def setCurrent(self, value: float) -> None:
        """Report a new illuminance reading. Clamped to [0, 200000] lx.

        Args:
            value: Illuminance reading in lux.

        Example:
            ```python
            illuminance.setCurrent(120)
            ```
        """
        self._write_state({IlluminanceProperty.Current.value: max(0, min(200000, value))})

    async def updateValue(self, property: str, value: Any) -> None:
        """Read-only sensor: external writes are ignored."""
